File: java/src/main/java/org/nc/knowledge/PartyType.java
package org.nc.knowledge;

public class PartyType {
    private final String name;
    private final PartyType supertype;

    public PartyType(String name) {
        this.name = name;
        this.supertype = null;
    }

    public PartyType(String name, PartyType supertype) {
        this.name = name;
        this.supertype = supertype;
    }

    public String getName() {
        return name;
    }

    public PartyType getSupertype() {
        return supertype;
    }

}

File: python/knowledge/PartyType.py
class PartyType:
    def __init__(self, name: str, supertype: "PartyType" = None):
        self.name = name
        self.supertype = supertype
    
    
    # (Derivazione) Restituisce l'insieme di tutti i tipi di party a cui appartiene il tipo corrente
    def get_all_types(self):
        # Essenzialmente risale la gerarchia dei tipi di partito fino a raggiungere il tipo radice
        types = [self]
        current = self.supertype
        while current:
            types.append(current)
            current = current.supertype
        return types


    def __repr__(self):
        return f"PartyType({self.name})"

